geth/peers/peerpool.go

```diff
@@ -92,42 +97,101 @@ func (p *PeerPool) Start(server *p2p.Server) error {
 	return nil
 }
 
+// restartDiscovery and search for topics that have peer count below min
+func (p *PeerPool) restartDiscovery(server *p2p.Server) error {
+	if server.DiscV5 == nil {
+		ntab, err := StartDiscv5(server)
+		if err != nil {
+			log.Error("starting discv5 failed", "error", err, "retry in", discoveryRestartTimeout)
+			return err
+		}
+		log.Debug("restarted discovery from peer pool")
+		server.DiscV5 = ntab
+	}
+	for _, t := range p.topics {
+		if !t.BelowMin() || t.SearchRunning() {
+			continue
+		}
+		err := t.StartSearch(server)
+		if err != nil {
+			log.Error("search failed to start", "error", err)
+		}
+	}
+	return nil
+}
+
 // handleServerPeers watches server peer events, notifies topic pools about changes
 // in the peer set and stops the discv5 if all topic pools collected enough peers.
 func (p *PeerPool) handleServerPeers(server *p2p.Server, events <-chan *p2p.PeerEvent) {
+	var (
+		retryDiscv5 <-chan time.Time
+	)
+
 	for {
 		select {
 		case <-p.quit:
 			return
+		case <-retryDiscv5:
+			if err := p.restartDiscovery(server); err != nil {
+				retryDiscv5 = time.After(discoveryRestartTimeout)
+			}
 		case event := <-events:
 			switch event.Type {
 			case p2p.PeerEventTypeDrop:
-				p.mu.Lock()
-				for _, t := range p.topics {
-					t.ConfirmDropped(server, event.Peer, event.Error)
-					// TODO(dshulyak) restart discv5 if peers number dropped too low
+				log.Debug("confirm peer dropped", "ID", event.Peer)
+				// if any peer is below min limit and search is not running
```

Comment:
kinda valid, but redundant, i think i will just drop it. maybe wording was bad though

p.stopOnMax is needed because we never stop search or discv5 on nodes without stopOnMax